datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1670,17 +1734,106 @@ fn inlist_except(mut l1: InList, l2: InList) -> Result<Expr> {
     Ok(Expr::InList(l1))
 }
 
+pub struct RewriteCycle {
+    consecutive_unchanged_count: usize,
+    total_iterations: usize,
+    num_rewriters: usize,
+}
+pub type RewriteCycleResult = ControlFlow<Result<Expr>, Expr>;
+
+impl RewriteCycle {
+    fn new() -> Self {
+        RewriteCycle {
+            // use usize::MAX as default to avoid checking null in is_done() comparison
```

Comment:
Is the reason to avoid checking null because of cost? I'm fine with either way, but curious on the choice.

Reply:
I think modeling this with `num_rewriters: Option<usize>` would make it clearer what is going on and help the compiler check the logic

Reply:
> I think modeling this with `num_rewriters: Option<usize>` would make it clearer what is going on and help the compiler check the logic

yes I tried to get too fancy here with micro-optimizations when I saw benchmark regressions 😆

Reply:
I think `cycle_length` would also be a more clear name for this variable 🤔  I think it is confusing that it is only initialized after the first iteration